pkg/controller/imagelistpulljob/imagelistpulljob_controller.go

```diff
@@ -0,0 +1,422 @@
+	// some images delete from ImageListPullJob.Spec.Images
+	for image, imagePullJob := range imagePullJobs {
+		if !slice.ContainsString(job.Spec.Images, image, nil) {
+			needToDelete = append(needToDelete, imagePullJob)
+		}
+	}
+	return needToCreate, needToDelete
+}
+
+func (r *ReconcileImageListPullJob) calculateStatus(job *appsv1alpha1.ImageListPullJob, imagePullJobs map[string]*appsv1alpha1.ImagePullJob) *appsv1alpha1.ImageListPullJobStatus {
+	var active, completed, succeeded int32
+	// record the failed image status
+	var failedImageStatuses []*appsv1alpha1.FailedImageStatus
+
+	for _, imagePullJob := range imagePullJobs {
+		if imagePullJob.Status.StartTime == nil {
+			continue
+		}
+
+		if imagePullJob.Status.Active > 0 {
+			active = active + 1
+		}
+
+		if imagePullJob.Status.Desired == (imagePullJob.Status.Failed + imagePullJob.Status.Succeeded) {
+			completed = completed + 1
+			//
+			if imagePullJob.Status.Failed > 0 {
+				failedImagePullJobStatus := &appsv1alpha1.FailedImageStatus{
+					ImagePullJob: imagePullJob.Name,
+					Name:         imagePullJob.Spec.Image,
+					Message:      fmt.Sprintf("Please check for details which nodes failed by using %s.", imagePullJob.Name),
+				}
+				failedImageStatuses = append(failedImageStatuses, failedImagePullJobStatus)
+			}
+
+		}
+
+		if imagePullJob.Status.Desired == imagePullJob.Status.Succeeded {
+			succeeded = succeeded + 1
+		}
+	}
+
+	// 4. status
+	newStatus := &appsv1alpha1.ImageListPullJobStatus{
+		Desired:             int32(len(job.Spec.Images)),
+		Active:              active,
+		Completed:           completed,
+		Succeeded:           succeeded,
+		StartTime:           job.Status.StartTime,
+		FailedImageStatuses: failedImageStatuses,
+	}
+
+	now := metav1.NewTime(r.clock.Now())
+	if newStatus.StartTime == nil {
+		newStatus.StartTime = &now
+	}
+
+	if job.Spec.CompletionPolicy.Type != appsv1alpha1.Never && newStatus.Desired == newStatus.Completed {
+		newStatus.CompletionTime = &now
+	}
+	return newStatus
+}
+
+func (r *ReconcileImageListPullJob) syncImagePullJob(job *appsv1alpha1.ImageListPullJob, needToCreate, needToDelete []*appsv1alpha1.ImagePullJob) error {
+
+	//1. manage creating
+	var errs []error
+	if len(needToCreate) > 0 {
+		var createdNum int
+		var createdErr error
+
+		createdNum, createdErr = util.SlowStartBatch(len(needToCreate), slowStartInitialBatchSize, func(idx int) error {
+			imagePullJob := needToCreate[idx]
+			key := types.NamespacedName{Namespace: job.Namespace, Name: job.Name}.String()
+			scaleExpectations.ExpectScale(key, expectations.Create, imagePullJob.Spec.Image)
+			err := r.Create(context.TODO(), imagePullJob)
+			if err != nil {
+				scaleExpectations.ObserveScale(key, expectations.Create, imagePullJob.Spec.Image)
+				if !errors.IsTimeout(err) {
+					return fmt.Errorf("fail to create ImagePullJob (%s/%s): %s", imagePullJob.Namespace, imagePullJob.Spec.Image, err.Error())
+				}
+			}
+
+			return nil
```

Comment:
I carefully looked through the implementation of SlowStartBatch and I believe that we should not ignore the case of errors and instead directly return err.